src/python/devcontainer-feature.json

```diff
@@ -25,13 +25,23 @@
     "installTools": {
       "type": "boolean",
       "default": true,
-      "description": "Install common Python tools like pylint"
+      "description": "Install common Python tools like pylint, override list using 'toolsToInstall'"
+    },
+    "toolsToInstall": {
```

Comment:
Looks like you are trying to do something similar to https://github.com/devcontainers/features/pull/768 which tries to add a new option `packages` with which users can install more utils with pip. 

Wondering if you want to sync this PR with https://github.com/devcontainers/features/pull/768 (this design was discussed with the broader team) or remove `toolsToInstall` changes as it would be conflicting/duplicate.

Reply:
When `installTools` is true, there is a default list of tools installed by PIPX. The `toolsToInstall` option changes this list to be whatever the container specifies.

I had an option similar to `packages` from #768, but opted not to include it for fear it would be controversial: I wanted to keep the base python free of non-standard libraries.

We can certainly pull in the `packages` option: it serves a different purpose to `toolsToInstall`.

Reply:
I'm working on pulling the changes in #768 into this PR, giving us both `toolsToInstall` and `packages`.

Reply:
> We can certainly pull in the packages option: it serves a different purpose to toolsToInstall.

Hmmm.. Is that because with `packages` option, we are installing with `pip`? and with `toolsToInstall` it's installing with `pip` in addition to overwriting pkgs installed with `installTools` ? 

How about `installTools` only installing new tools with pipx? If user don't want to install pkgs that gets installed with `installTools`, they can simply add `installTools`: "false"

Reply:
`installTools` installs tools (command line utilities) via `pipx`. Originally `installTools` was a boolean that triggered installation of a fixed-set of additional, to python, tools. It does not affect the installed python in any way.

We use a different set of tools, for example the default set includes `pylint`, we use `ruff`.

I added `toolsToInstall`, with a default value of the list of tools that `installTools` already installed, but giving us a way to override the default set. This could be simplified by making `installTools` a string that is a comma-separated list of tools to install, with an empty list resulting in no tools being installed via `pipx`. This would get rid of the need for `toolsToInstall`.

Meanwhile, `packages` uses `pip` to install python packages into the python system that is installed by this feature. It "pollutes" the installed python system. That is why issues such as #577 arise. In fact, I suspect that the `packages` option will cause the same error as seen in #577 when `version` is set to `os-provided`, and Debian `bookworm` is the base image. I'll be testing that and see if it can be worked around.

Reply:
I just remembered: converting `toolsToInstall` to a string will break existing usage of the feature. This would either require a major revision bump, or leaving it a `boolean` and implementing it the way I did to allow for backwards comaptibility.

Reply:
> Hmmm.. Is that because with packages option, we are installing with pip? and with toolsToInstall it's installing with pip in addition to overwriting pkgs installed with installTools ?

With `installTools` set to `true`, `pip` is used to bootstrap an installation of `pipx` in a way that does not affect the installed python (note: if the installed python has a managed environment setting, then `pipx` is installed using the package manager and this does affect the installed python, but in an acceptable manner). These tools are expected to include a command line interface: they are callable via scripts and users on the command line (shell).

The behavior of `installTools` and `toolsToInstall` is as follows:

| installTools | toolsToInstall | Result | Effect on Installed Python |
---------------------------------------------------------------
| false | value is ignored | no addtional tools are installed. | None |
| true | unset | default set of tools is installed | None |
| true | `"tool_one,tool_two"` | default set of tools **not** installed: `tool_one` and `tool_two` are installed instead. | None |

Meanwhile `packages` uses `pip` to install packages directly into the installed python. If this package does not include a command line interface, it makes no sense to install it via `pipx`: doing so will not make the package available to the installed python. Because using `pip` to install packages causes the installed python to be modified, this will cause the issues seen in #577 if we're not careful. It also makes it likely that developers will miss dependencies they should include in their package definition because they are already provided by they installed python.

I had included a `packages` like capability, then removed it because of the dependency issue/concern. I had concluded that a 'local-feature' could always add pip installed packages if desired.

For now, I've added, and am testing, having `packages` in then python feature. It will be limited for some cases, e.g. `version="os-provided"` and python is a managed environment: in that case version specifiers on the pip installed package will be ignored.

Reply:
Got it, thanks for the detailed explanation.

<img width="935" alt="image" src="https://github.com/devcontainers/features/assets/24955913/5dc54dc6-3359-45f2-bb62-69244f96af7c">

(Apologies if I am misunderstanding) Why do we ignore `toolsToInstall` if `installTools` is false? I am just worried if it might be confusing to folks 🤔 

> With installTools set to true, pip is used to bootstrap an installation of pipx in a way that does not affect the installed python (note: if the installed python has a managed environment setting, then pipx is installed using the package manager and this does affect the installed python, but in an acceptable manner). These tools are expected to include a command line interface: they are callable via scripts and users on the command line (shell).

Instead of ignoring, we could installs tools similar to ^

Reply:
> . This would either require a major revision bump, or leaving it a boolean and implementing it the way I did to allow for backwards comaptibility.

I think we should avoid bumping major versions unless it's completely necessary. 😅

Reply:
> Apologies if I am misunderstanding) Why do we ignore toolsToInstall if installTools is false? I am just worried if it might be confusing to folks 🤔

`installTools` is a boolean flag that says whether or not to install the tools listed in `toolsToInstall`.

I implemented it this way to maintain backward compatibility with the existing `python` feature behavior.

If we make `installTools` a string, then we break backwards compatibility, but get a single variable that lists the tools to install. If empty no tools would be installed. Basically, I prioritized backwards compatibility.

Perhaps the answer to confusion is to make the definitions/descriptions read as follows:

```
"installTools": {
      "type": "boolean",
      "default": true,
      "description": "Flag indicating whether or not to install the tools specified via the 'toolsToInstall' option. Default is 'true'."
    },
    "toolsToInstall": {
      "type": "string",
      "default": "flake8,autopep8,black,yapf,mypy,pydocstyle,pycodestyle,bandit,pipenv,virtualenv,pytest,pylint",
      "description": "Comma-separated list of tools to install when installTools is true. Defaults to a set of common Python tools like pylint."
    },
```

Reply:
Commit [baf1568](https://github.com/devcontainers/features/pull/830/commits/baf1568d80bb6f1eb73dd7e1b774b2fa5f3b662d) addresses this.